pkg/util/pki/csr.go

```diff
@@ -27,24 +27,14 @@ import (
 	"fmt"
 	"math/big"
 	"net"
+	"net/url"
+	"strings"
 	"time"
 
 	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
 	"github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
 )
 
-// CommonNameForCertificate returns the common name that should be used for the
-// given Certificate resource, by inspecting the CommonName and DNSNames fields.
-func CommonNameForCertificate(crt *v1alpha2.Certificate) string {
-	if crt.Spec.CommonName != "" {
-		return crt.Spec.CommonName
-	}
-	if len(crt.Spec.DNSNames) == 0 {
-		return ""
-	}
-	return crt.Spec.DNSNames[0]
-}
-
 // DNSNamesForCertificate returns the DNS names that should be used for the
 // given Certificate resource, by inspecting the CommonName and DNSNames fields.
 func DNSNamesForCertificate(crt *v1alpha2.Certificate) []string {
```

Comment:
Should we automatically add the provided commonName to the dnsNames? I feel like we shouldn't - if a user specifies:

```
commonName: example.com
dnsNames:
- www.example.com
```

We should accept a certificate for

```
commonName: example.com
dnsNames:
- example.com
- www.example.com
```

as valid (despite requesting just one dnsName):

```
commonName: example.com
dnsNames:
- www.example.com
```

Reply:
i.e. I'd expect this PR to mostly do two things:

1) stop all magic juggling between CN and dnsNames in our *request* flow
2) make our 'certificate is up to date' checking more permissive between the CN and dnsNames fields

Reply:
:+1: